doom/config.org
:DOC-CONFIG:
#+property: header-args:emacs-lisp :tangle config.el :mkdirp yes :comments no
#+startup: fold
:END:
* Guidance
 Whenever you reconfigure a package, make sure to wrap your config in an
 `after!' block, otherwise Doom's defaults may override your settings. E.g.

   (after! PACKAGE
     (setq x y))

 The exceptions to this rule:

   - Setting file/directory variables (like `org-directory')
   - Setting variables which explicitly tell you to set them before their
     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
   - Setting doom variables (which start with 'doom-' or '+').

 Here are some additional functions/macros that will help you configure Doom.

 - `load!' for loading external *.el files relative to this one
 - `use-package!' for configuring packages
 - `after!' for running code after a package has loaded
 - `add-load-path!' for adding directories to the `load-path', relative to
   this file. Emacs searches the `load-path' when you load packages with
   `require' or `use-package'.
 - `map!' for binding new keys

 To get information about any of these functions/macros, move the cursor over
 the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
 This will open documentation for it, including demos of how they are used.
 Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
 etc).

 You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
 they are implemented.

* Automatic config
** Environment variables
#+begin_src emacs-lisp :tangle yes
(use-package! dotenv
  :config
  ;; Load .env from ~/code/github.com/dotfiles/.env on startup
  (dotenv-update-project-env "~/code/github.com/dotfiles/")

  (add-hook 'projectile-after-switch-project-hook
            (lambda ()
              "Loading project-specific .env..."
              (dotenv-update-project-env (projectile-project-root)))))
#+end_src

*** Confirm the envars are being loaded
#+begin_src emacs-lisp
;; (getenv "GPTEL_API_KEY")
#+end_src

** Coding
*** Java
**** Lsp
#+begin_src emacs-lisp :tangle yes
(add-hook! 'java-mode #'lsp)
(add-hook! 'lsp-mode-hook #'lsp-lens-mode)
(add-hook! 'java-mode-hook #'lsp-java-boot-lens-mode)
(add-hook! 'java-mode-hook #'lsp-deferred)
#+end_src
**** Debugging
#+begin_src emacs-lisp :tangle yes
(map! :leader "d t" #'dap-breakpoint-toggle :desc "Debugger toggle breakpoint")
(map! :leader "d s" #'dap-debug :desc "Debugger start")
(map! :leader "d r" #'dap-debug-restart :desc "Debugger restart")
(map! :leader "d j" #'dap-java-debug-test-class :desc "Debug JUnit test")
(map! :leader "d d" #'dap-disconnect :desc "Debugger disconnect")
(map! :leader "d n" #'dap-next :desc "Debugger next")
(map! :leader "d o" #'dap-step-out :desc "Debugger step over")
(map! :leader "d i" #'dap-step-in :desc "Debugger step in")
(map! :leader "d l" #'dap-locals :desc "Debugger locals")
(map! :leader "d e a" #'dap-ui-expressions-add :desc "Debugger Expression Add")
(map! :leader "d e l" #'dap-ui-expressions :desc "Debugger Expression list")
(map! :leader "d e r" #'dap-ui-expressions-remove :desc "Debugger Expression Remove")
(map! :leader "d b b" #'dap-ui-breakpoints-browse :desc "Debugger Breakpoints Browse")
(map! :leader "d b c" #'dap-breakpoint-delete-all :desc "Debugger Breakpoint Clear")
#+end_src

**** Custom workflow improvements
***** Run current file
Runs the file that is currently open in the buffer in an Eshell instance
#+begin_src emacs-lisp :tangle yes
(defun java-eshell()
  (interactive)
  (let ((file buffer-file-name)
        (buf (get-buffer-create "*eshell-java*")))
    (pop-to-buffer buf)
    (unless (eq major-mode 'eshell-mode)
      (eshell-mode))
    ;; move to end before inserting
    (goto-char (point-max))
    (insert (format "java %s" file))
    (eshell-send-input)))

(map! :leader "j e" #'java-eshell :desc "[J]ava [e]xecute code")
#+end_src

*** Node
**** Nvm
#+begin_src emacs-lisp :tangle yes
(setenv "NVM_DIR" (expand-file-name "~/.nvm"))
(let ((node-path (concat (getenv "NVM_DIR") "/versions/node/v25.2.1/bin")))
  (setenv "PATH" (concat node-path ":" (getenv "PATH")))
  (setq exec-path (cons node-path exec-path)))
#+end_src
*** TypeScript
**** LSP
#+begin_src emacs-lisp :tangle yes
(add-to-list 'auto-mode-alist '("\\.tsx\\'" . web-mode))
(add-hook! 'web-mode-hook #'lsp)
#+end_src

** Eshell
#+begin_src emacs-lisp :tangle yes
(add-hook! 'eshell-first-time-mode-hook
           (eshell/alias "jenv" "~/.jenv/bin/jenv $*"))
#+end_src
** Default shell
#+begin_src emacs-lisp :tangle yes
(setq vterm-shell "/usr/bin/zsh")
#+end_src

#+RESULTS:
: /usr/bin/zsh

** Theme settings
*** Theme color scheme
#+begin_src emacs-lisp :tangle yes
(setq doom-theme 'doom-one)
#+end_src

*** Display line numbers
#+begin_src emacs-lisp :tangle yes
(setq display-line-numbers-type 'relative)
#+end_src
*** Default window size
#+begin_src emacs-lisp :tangle yes
(modify-frame-parameters nil '((undecorated . t) (skip-taskbar . t)))
#+end_src

** Emacs everywhere

#+begin_src emacs-lisp :tangle yes
(setq emacs-everywhere-frame-parameters
      '((width . 80)        ;; ~600px)
        (height . 15)       ;; 15 characters tall
        (undecorated . t)   ;; remove title bar
        (skip-taskbar . t)  ;; do not show in taskbar
        (vertical-scroll-bars . nil)
        (horizontal-scroll-bars . nil)))
#+end_src

** Org mode settings
*** General
#+begin_src emacs-lisp :tangle yes
(setq org-directory "~/org/")

;; Keymaps
(map! :leader (:prefix "t" :desc "[T]angle [O]rg file" "o" 'org-babel-tangle))
#+end_src
**** Fold drawers by default
Ensures the targeted drawers are always collapsed when opening the file.
#+begin_src emacs-lisp :tangle yes :results none
(setq org-drawers '("PROPERTIES" "LOGBOOK"))

(defun org-refold-drawers()
  (save-excursion
    (goto-char (point-min))
    (while (re-search-forward "^:\\(PROPERTIES\\|LOGBOOK\\):" nil t)
      (org-flag-drawer t))))

(add-hook! 'org-cycle-hook 'org-refold-drawers)
#+end_src

*** Anki
#+begin_src emacs-lisp :tangle yes
(use-package! anki-editor)

;; Keymaps
(map! :leader "n p" 'anki-editor-push-notes :desc "Push Anki notes" )
#+end_src

** Music
*** eradio
**** Startup
This autoplays the radio when emacs starts up for the first time.
#+begin_src emacs-lisp :tangle yes
;; (eradio-play "https://somafm.com/bossa256.pls")
#+end_src

**** Keymaps
#+begin_src emacs-lisp :tangle yes
(map! :leader (:prefix ("r" . "eradio") :desc "Play a radio channel" "p" 'eradio-play))
(map! :leader (:prefix ("r" . "eradio") :desc "Stop the radio player" "s" 'eradio-stop))
(map! :leader (:prefix ("r" . "eradio") :desc "Toggle the radio player" "t" 'eradio-toggle))
#+end_src

**** Radio list
#+begin_src emacs-lisp :tangle yes
(setq eradio-channels '(("def con - soma fm" . "https://somafm.com/defcon256.pls")         ;; electronica with defcon-speaker bumpers
                        ("metal - soma fm"   . "https://somafm.com/metal130.pls")          ;; \m/
                        ("bossa beyond - soma fm"  . "https://somafm.com/bossa256.pls")    ;; bossa nova
                        ("ambient - soma fm"  . "https://somafm.com/groovesalad256.pls")   ;; ambient and chill
                        ("specials - soma fm"  . "https://somafm.com/nossl/specials.pls")   ;; Soma fm specials, mixed
                        ))
#+end_src

** Search
*** Web
Leverages =counsel-web-search= to open a mini-buffer at the bottom of the screen, showing the top results of the search

#+begin_src emacs-lisp :tangle yes
(map! :leader "s w" 'eww :desc "Search the web")
(setq eww-auto-rename-buffer 'title)
(setq eww-buffers-queryable nil)
#+end_src
*** Docs
#+begin_src emacs-lisp :tangle yes
(map! :leader "s d" 'devdocs-lookup :desc "Search offline docs")
#+end_src
** AI
*** gptel
#+begin_src emacs-lisp :tangle yes
(use-package! gptel)
(setq gptel-default-mode 'org-mode)
(setq
 gptel-model 'gemini-flash-latest
 gptel-backend (gptel-make-gemini "Gemini"
                 :key (getenv "GPTEL_API_KEY")
                 :stream t))
#+end_src

** Discord
*** Startup
Install package to share status in emacs using Discord Rich Presence
#+begin_src emacs-lisp :tangle yes
(setq elcord-editor-icon "emacs_material_icon")
(elcord-mode 1)
#+end_src

* Manual settings
** Swapping caps for escape
This needs to be done in $HOME/.profile: =/usr/bin/setxkbmap -option "caps:swapescape"=
Or C-c C-c in the code block below:
#+begin_src bash
echo /usr/bin/setxkbmap -option "caps:swapescape" >> $HOME/.profile
#+end_src

Alternatively, set up the startup application in your Distro settings:
#+begin_src bash
/usr/bin/setxkbmap -option "caps:swapescape"
#+end_src

#+RESULTS:

** Emacs everywhere
This needs to be set to a keybind in your system
#+begin_src bash
emacsclient --eval "(emacs-everywhere)"
#+end_src

** AnkiConnect
Install the AnkiConnect addon.
*Code:* 2055492159
** Setting up forge
1. Create a personal token in GitHub with Issue and Pull Request permissions
2. Create =~/.authinfo=
3. Add the following to the file: =machine api.github.com login <YOUR_GITHUB_USERNAME>^forge password <YOUR_GITHUB_TOKEN>=


doom/config.el
(use-package! dotenv
  :config
  ;; Load .env from ~/code/github.com/dotfiles/.env on startup
  (dotenv-update-project-env "~/code/github.com/dotfiles/")

  (add-hook 'projectile-after-switch-project-hook
            (lambda ()
              "Loading project-specific .env..."
              (dotenv-update-project-env (projectile-project-root)))))

;; (getenv "GPTEL_API_KEY")

(add-hook! 'java-mode #'lsp)
(add-hook! 'lsp-mode-hook #'lsp-lens-mode)
(add-hook! 'java-mode-hook #'lsp-java-boot-lens-mode)
(add-hook! 'java-mode-hook #'lsp-deferred)

(map! :leader "d t" #'dap-breakpoint-toggle :desc "Debugger toggle breakpoint")
(map! :leader "d s" #'dap-debug :desc "Debugger start")
(map! :leader "d r" #'dap-debug-restart :desc "Debugger restart")
(map! :leader "d j" #'dap-java-debug-test-class :desc "Debug JUnit test")
(map! :leader "d d" #'dap-disconnect :desc "Debugger disconnect")
(map! :leader "d n" #'dap-next :desc "Debugger next")
(map! :leader "d o" #'dap-step-out :desc "Debugger step over")
(map! :leader "d i" #'dap-step-in :desc "Debugger step in")
(map! :leader "d l" #'dap-locals :desc "Debugger locals")
(map! :leader "d e a" #'dap-ui-expressions-add :desc "Debugger Expression Add")
(map! :leader "d e l" #'dap-ui-expressions :desc "Debugger Expression list")
(map! :leader "d e r" #'dap-ui-expressions-remove :desc "Debugger Expression Remove")
(map! :leader "d b b" #'dap-ui-breakpoints-browse :desc "Debugger Breakpoints Browse")
(map! :leader "d b c" #'dap-breakpoint-delete-all :desc "Debugger Breakpoint Clear")

(defun java-eshell()
  (interactive)
  (let ((file buffer-file-name)
        (buf (get-buffer-create "*eshell-java*")))
    (pop-to-buffer buf)
    (unless (eq major-mode 'eshell-mode)
      (eshell-mode))
    ;; move to end before inserting
    (goto-char (point-max))
    (insert (format "java %s" file))
    (eshell-send-input)))

(map! :leader "j e" #'java-eshell :desc "[J]ava [e]xecute code")

(setenv "NVM_DIR" (expand-file-name "~/.nvm"))
(let ((node-path (concat (getenv "NVM_DIR") "/versions/node/v25.2.1/bin")))
  (setenv "PATH" (concat node-path ":" (getenv "PATH")))
  (setq exec-path (cons node-path exec-path)))

(add-to-list 'auto-mode-alist '("\\.tsx\\'" . web-mode))
(add-hook! 'web-mode-hook #'lsp)

(add-hook! 'eshell-first-time-mode-hook
           (eshell/alias "jenv" "~/.jenv/bin/jenv $*"))

(setq vterm-shell "/usr/bin/zsh")

(setq doom-theme 'doom-one)

(setq display-line-numbers-type 'relative)

(modify-frame-parameters nil '((undecorated . t) (skip-taskbar . t)))

(setq emacs-everywhere-frame-parameters
      '((width . 80)        ;; ~600px)
        (height . 15)       ;; 15 characters tall
        (undecorated . t)   ;; remove title bar
        (skip-taskbar . t)  ;; do not show in taskbar
        (vertical-scroll-bars . nil)
        (horizontal-scroll-bars . nil)))

(setq org-directory "~/org/")

;; Keymaps
(map! :leader (:prefix "t" :desc "[T]angle [O]rg file" "o" 'org-babel-tangle))

(setq org-drawers '("PROPERTIES" "LOGBOOK"))

(defun org-refold-drawers()
  (save-excursion
    (goto-char (point-min))
    (while (re-search-forward "^:\\(PROPERTIES\\|LOGBOOK\\):" nil t)
      (org-flag-drawer t))))

(add-hook! 'org-cycle-hook 'org-refold-drawers)

(use-package! anki-editor)

;; Keymaps
(map! :leader "n p" 'anki-editor-push-notes :desc "Push Anki notes" )

;; (eradio-play "https://somafm.com/bossa256.pls")

(map! :leader (:prefix ("r" . "eradio") :desc "Play a radio channel" "p" 'eradio-play))
(map! :leader (:prefix ("r" . "eradio") :desc "Stop the radio player" "s" 'eradio-stop))
(map! :leader (:prefix ("r" . "eradio") :desc "Toggle the radio player" "t" 'eradio-toggle))

(setq eradio-channels '(("def con - soma fm" . "https://somafm.com/defcon256.pls")         ;; electronica with defcon-speaker bumpers
                        ("metal - soma fm"   . "https://somafm.com/metal130.pls")          ;; \m/
                        ("bossa beyond - soma fm"  . "https://somafm.com/bossa256.pls")    ;; bossa nova
                        ("ambient - soma fm"  . "https://somafm.com/groovesalad256.pls")   ;; ambient and chill
                        ("specials - soma fm"  . "https://somafm.com/nossl/specials.pls")   ;; Soma fm specials, mixed
                        ))

(map! :leader "s w" 'eww :desc "Search the web")
(setq eww-auto-rename-buffer 'title)
(setq eww-buffers-queryable nil)

(map! :leader "s d" 'devdocs-lookup :desc "Search offline docs")

(use-package! gptel)
(setq gptel-default-mode 'org-mode)
(setq
 gptel-model 'gemini-flash-latest
 gptel-backend (gptel-make-gemini "Gemini"
                 :key (getenv "GPTEL_API_KEY")
                 :stream t))

(setq elcord-editor-icon "emacs_material_icon")
(elcord-mode 1)
